### PR TITLE
ddl: Allow global Index including all columns in the partitioning expression

### LINKED ISSUE #56230: global index limitation: Unsupported Global Index including all columns in the partitioning expression
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

We are encountering hotspots on both the AUTO_INCREMENT primary key and the updated_at index during ingestion. To address this, I plan to implement hash/key partitioning to alleviate the hotspots on both the primary key and the secondary index (idx_updated_at). The specific actions I intend to take are:

1. Keep the primary key and idx_updated_at as local indexes to distribute load and mitigate the hotspots.
2. Convert the index on user_id (idx_user_id) into a global index. Since user_id is not unique, I will modify the index by including the id column, ensuring idx_user_id becomes a unique index.

```
# original schema
CREATE table t (
  id bigint NOT NULL AUTO_INCREMENT,
  updated_at timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  user_id bigint DEFAULT NULL,
  PRIMARY KEY (id),
  KEY idx_user_id (user_id, id),
  KEY idx_updated_at (updated_at)
) ENGINE=InnoDB DEFAULT CHARSET=latin1;

set session tidb_enable_global_index=on;

CREATE table t (
  id bigint NOT NULL AUTO_INCREMENT,
  updated_at timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  user_id bigint DEFAULT NULL,
  PRIMARY KEY (id),
  UNIQUE KEY idx_user_id (user_id, id) global,
  KEY idx_updated_at (updated_at)
) ENGINE=InnoDB DEFAULT CHARSET=latin1
PARTITION BY KEY(id)
PARTITIONS 64;
```

### 2. What did you expect to see? (Required)
no error

### 3. What did you see instead (Required)
```
mysql> CREATE table t (
    ->   id bigint NOT NULL AUTO_INCREMENT,
    ->   updated_at timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
    ->   user_id bigint DEFAULT NULL,
    ->   PRIMARY KEY (id),
    ->   UNIQUE KEY idx_user_id (user_id, id) global,
    ->   KEY idx_updated_at (updated_at)
    -> ) ENGINE=InnoDB DEFAULT CHARSET=latin1
    -> PARTITION BY KEY(id)
    -> PARTITIONS 64;
ERROR 8200 (HY000): Unsupported Global Index including all columns in the partitioning expression
```

### 4. What is your TiDB version? (Required)

v8.3.0

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #56230

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Allow global Index including all columns in the partitioning expression.
```

### CODE REVIEW

File: pkg/ddl/executor.go [hunk 1]
```diff
@@ -4515,6 +4506,28 @@ func GetName4AnonymousIndex(t table.Table, colName pmodel.CIStr, idxName pmodel.
 	return indexName
 }
 
+func checkCreateUniqueGlobalIndex(ec errctx.Context, tblInfo *model.TableInfo, indexName string, indexColumns []*model.IndexColumn, isUnique bool, isGlobal bool) error {
+	if isGlobal && tblInfo.GetPartitionInfo() == nil {
```
Comment (mjonss) [inline pkg/ddl/executor.go:4510]: Maybe use a helper variable to make it more readable, and use it for the whole function?
```suggestion
	pi := tblInfo.GetPartitionInfo()
	if isGlobal && pi == nil {
```

File: tests/integrationtest/r/globalindex/mem_index_lookup.result [hunk 2]
```diff
@@ -55,6 +56,39 @@ select * from t partition(p0, p1) use index(idx1) where b <= 2;
 a	b
 1	2
 5	1
+explain select * from t use index(idx) where a > 2;
+id	estRows	task	access object	operator info
+Projection_5	3333.33	root		globalindex__mem_index_lookup.t.a, globalindex__mem_index_lookup.t.b
+└─UnionScan_6	3333.33	root		gt(globalindex__mem_index_lookup.t.a, 2)
+  └─IndexLookUp_10	3333.33	root	partition:all
```
Comment (mjonss) [inline tests/integrationtest/r/globalindex/mem_index_lookup.result:63]: Should we have the `partition:...` clause in the `access object` for global indexes?

**Review by mjonss — COMMENTED** [review]
LGTM, with some minor questions

**Review by mjonss — APPROVED** [review]

File: pkg/ddl/executor.go [hunk 3]
```diff
@@ -4515,6 +4506,28 @@ func GetName4AnonymousIndex(t table.Table, colName pmodel.CIStr, idxName pmodel.
...
+		return dbterror.ErrGeneralUnsupportedDDL.GenWithStackByArgs("Global Index on non-partitioned table")
+	}
+	if isGlobal && !isUnique {
+		// TODO: remove this limitation
+		return dbterror.ErrGeneralUnsupportedDDL.GenWithStackByArgs("Global IndexOption on non-unique index")
+	}
+	if isUnique && tblInfo.GetPartitionInfo() != nil {
```
Comment (Defined2014) [inline pkg/ddl/executor.go:4517]:
```suggestion
	if isUnique && pi != nil {
```

File: tests/integrationtest/r/globalindex/mem_index_lookup.result (hunk 2 above)
Comment (Defined2014) [reply tests/integrationtest/r/globalindex/mem_index_lookup.result:63]:
```
mysql> explain select b from t use index(idx) where a < 10;
+---------------------------------+---------+-----------+-----------------------+-------------------------------------------------+
| id                              | estRows | task      | access object         | operator info                                   |
+---------------------------------+---------+-----------+-----------------------+-------------------------------------------------+
| Projection_4                    | 3323.33 | root      |                       | test.t.b                                        |
| └─IndexLookUp_10                | 3323.33 | root      | partition:p0          |                                                 |
|   ├─Selection_9(Build)          | 3323.33 | cop[tikv] |                       | in(_tidb_tid, 514)                              |
|   │ └─IndexRangeScan_6          | 3323.33 | cop[tikv] | table:t, index:idx(a) | range:[-inf,10), keep order:false, stats:pseudo |
|   └─TableRowIDScan_7(Probe)     | 3323.33 | cop[tikv] | table:t               | keep order:false, stats:pseudo                  |
+---------------------------------+---------+-----------+-----------------------+-------------------------------------------------+
5 rows in set (0.00 sec)
```

We could get prune information from `access object` and `Selection`. And I think `access object` is more intuitive.

**Review by tangenta — APPROVED** [review]

**Pull request merged**